File: src/Cocoar.SignalARRR.Swift/Tests/CocoarSignalARRRIntegrationTests/ErrorHandlingTests.swift
import Foundation
import XCTest
@testable import CocoarSignalARRR

final class ErrorHandlingTests: IntegrationTestBase {

    func testStructuredError_ArgumentException() async throws {
        do {
            let _: String = try await connection.invoke(
                "ExtraMethods.ThrowArgumentException",
                arguments: "testParam"
            )
            XCTFail("Expected an error to be thrown")
        } catch {
            let harrrError = parseHARRRError(fromMessage: "\(error)")
            XCTAssertEqual(harrrError.type, "System.ArgumentException")
            XCTAssert(harrrError.message.contains("Invalid value provided"),
                      "Expected message to contain 'Invalid value provided', got: \(harrrError.message)")
        }
    }

    /// Contrast with the ArgumentException case above, which still arrives verbatim: that one names
    /// a pipeline stage the server controls, this one is whatever the hub method threw and could
    /// say anything about the server's insides. Since 5.0 it is withheld and logged server-side
    /// under the correlation id the client is shown.
    func testStructuredError_UnexpectedException_WithholdsTheDetail() async throws {
        do {
            let _: String = try await connection.invoke(
                "ExtraMethods.ThrowInvalidOperation"
            )
            XCTFail("Expected an error to be thrown")
        } catch {
            let harrrError = parseHARRRError(fromMessage: "\(error)")
            XCTAssertNotEqual(harrrError.type, "System.InvalidOperationException")
            XCTAssertFalse(harrrError.message.contains("This operation is not allowed"),
                           "The method's own message must not reach the client, got: \(harrrError.message)")
            XCTAssert(harrrError.message.contains("Correlation id:"),
                      "Expected a correlation id to trace this call by, got: \(harrrError.message)")
        }
    }

    func testInvokeNonExistentMethod_Throws() async throws {
        do {
            let _: String = try await connection.invoke("NonExistentMethod")
            XCTFail("Expected an error to be thrown")
        } catch {
            // Any error is acceptable — the key is it doesn't succeed
            XCTAssertFalse("\(error)".isEmpty)
        }
    }

    // MARK: - Error codes

    func testApplicationCodeTravelsVerbatim() async throws {
        do {
            let _: String = try await connection.invoke("ExtraMethods.ThrowRoomFull")
            XCTFail("Expected an error to be thrown")
        } catch {
            let harrrError = parseHARRRError(fromMessage: "\(error)")
            XCTAssertEqual(harrrError.code, "room_full")
            XCTAssertEqual(harrrError.normalizedCode, HARRRErrorCodes.internal, "an unknown code folds to internal")
            XCTAssertEqual(harrrError.message, "The room is full.")
        }
    }

    func testUnknownMethodIsMethodNotFound() async throws {
        do {
            let _: String = try await connection.invoke("NonExistentMethod")
            XCTFail("Expected an error to be thrown")
        } catch {
            XCTAssertEqual(parseHARRRError(fromMessage: "\(error)").normalizedCode, HARRRErrorCodes.methodNotFound)
        }
    }

    func testWrongArgumentCountIsInvalidArgumentCount() async throws {
        do {
            let _: String = try await connection.invoke("Echo", arguments: "one", "two")
            XCTFail("Expected an error to be thrown")
        } catch {
            XCTAssertEqual(parseHARRRError(fromMessage: "\(error)").normalizedCode, HARRRErrorCodes.invalidArgumentCount)
        }
    }
}

File: src/Cocoar.SignalARRR.Kotlin/signalarrr-integration-tests/src/test/kotlin/dev/cocoar/signalarrr/integration/ErrorHandlingTests.kt
package dev.cocoar.signalarrr.integration

import dev.cocoar.signalarrr.HARRRErrorCodes
import dev.cocoar.signalarrr.HARRRException
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNotEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows

class ErrorHandlingTests : IntegrationTestBase() {

    @Test
    fun `ArgumentException arrives structured and verbatim`() = test {
        val e = assertThrows<HARRRException> {
            connection.invoke<String>("ExtraMethods.ThrowArgumentException", "testParam")
        }
        assertEquals(HARRRErrorCodes.ARGUMENT_BINDING_FAILED, e.code)
        assertEquals("System.ArgumentException", e.error.type)
        assertTrue(e.error.message.contains("Invalid value provided"), e.error.message)
    }

    /**
     * Since server 5.0 an unexpected exception is withheld: the client sees a fixed sentence and a
     * correlation id, never the method's own message.
     */
    @Test
    fun `unexpected exception withholds the detail`() = test {
        val e = assertThrows<HARRRException> {
            connection.invoke<String>("ExtraMethods.ThrowInvalidOperation")
        }
        assertEquals(HARRRErrorCodes.INTERNAL, e.code)
        assertNotEquals("System.InvalidOperationException", e.error.type)
        assertFalse(e.error.message.contains("This operation is not allowed"), e.error.message)
        assertTrue(e.error.message.contains("Correlation id:"), e.error.message)
    }

    @Test
    fun `application error code travels verbatim`() = test {
        val e = assertThrows<HARRRException> {
            connection.invoke<String>("ExtraMethods.ThrowRoomFull")
        }
        assertEquals("room_full", e.error.code)
        assertEquals(HARRRErrorCodes.INTERNAL, e.code, "an unknown code folds to internal")
        assertEquals("The room is full.", e.error.message)
    }

    @Test
    fun `unknown method is method_not_found`() = test {
        val e = assertThrows<HARRRException> {
            connection.invoke<String>("NonExistentMethod")
        }
        assertEquals(HARRRErrorCodes.METHOD_NOT_FOUND, e.code)
    }

    @Test
    fun `wrong argument count is reported`() = test {
        val e = assertThrows<HARRRException> {
            connection.invoke<String>("Echo", "one", "two")
        }
        assertEquals(HARRRErrorCodes.INVALID_ARGUMENT_COUNT, e.code)
    }

    @Test
    fun `stream error surfaces as HARRRException`() = test {
        val e = assertThrows<HARRRException> {
            connection.stream<Int>("NonExistentStream").collect { }
        }
        assertEquals(HARRRErrorCodes.METHOD_NOT_FOUND, e.code)
    }
}
